Add Vertex AI express mode support

NijikaMyWaifu opened the pull request:
https://cloud.google.com/vertex-ai/generative-ai/docs/start/express-mode/overview

<!-- Put X in the box below to confirm -->

## Checklist:

- [X] I have read the [Contributing guidelines](https://github.com/SillyTavern/SillyTavern/blob/release/CONTRIBUTING.md).

Cohee1207 commented:
Very cool, thanks for the contribution.
I see in the docs they have different lists of supported models, and likely different API capabilities. You also can't reuse API keys between them. How about splitting these two API types instead of reusing AI Studio? That will be clearer for the end user.

NijikaMyWaifu committed:
Add Vertex AI express mode support
Split Google AI Studio and Vertex AI

NijikaMyWaifu commented:
Considering potential future support for the full version of Vertex AI, I've included all models in the model list here.

And because AI Studio API and Vertex AI API are basically the same, I've chosen to share most of the code here.

Cohee1207 commented:
Google made is unnecessarily complicated to get an API key for this mode, I'm still struggling with that. Thinking that would be common among other users too. I also think that calling it just Vertex would not set correct expectations. People would try to use their regular vertex keys and they will be rejected. So far Oauth flow is the biggest obstacle, but we can do it for OpenRouter, so it should be possible to get proper vertex integrated one day.

NijikaMyWaifu commented:
To use the full version of Vertex AI, a Service Account .json file is required. SillyTavern would then use this file to obtain a temporary Access Token and add it to the Bearer Header for API access. Perhaps we can first support the express mode and add support for the full version of Vertex AI later.
I also found that using location-specific domains like us-central1-aiplatform.google.com is much faster than aiplatform.google.com (up to 10 seconds faster).

AliceRabbit commented:
I believe it's reasonable to launch support for the more easily implementable Vertex Express mode before full Vertex AI support, especially given the readily available free trial for Express mode. I'm wondering if this feature will be merged into the code soon. 🙂

Cohee1207 reviewed (changes requested):
1. It doesn't specify anywhere that it only accepts Express mode keys. This should be specified both in dropdown and on form. This is most important.
2. It is missing every sampler setting, as well as other source specific settings such as system prompt toggle.
3. The list of models doesn't reflect the list of models supported in the Express mode. There're only these four: gemini-2.0-flash-001, gemini-2.0-flash-lite-001, gemini-2.5-pro-preview-03-25, gemini-2.5-flash-preview-04-17. They do not offer any gemma or gemini 1.5 models. And likely won't.
4. What about features like tool calling and image inlining? Both would not work currently.
5. The code has changes to the caption endpoint, but doesn't have relevant changes to the caption extension. Captioning can be left out for now.
6. Current code wouldn't read `vertexai_model` field from presets.

NijikaMyWaifu committed:
Merge branch 'staging' into vertexai

# Conflicts:
#	public/scripts/openai.js

NijikaMyWaifu committed:
Add support for Vertex AI, including updating default models and related settings, modifying frontend HTML to include Vertex AI options, and adjusting request processing logic in the backend API.

NijikaMyWaifu commented:
> 1. It doesn't specify anywhere that it only accepts Express mode keys. This should be specified both in dropdown and on form. This is most important.

Fixed

> 2. It is missing every sampler setting, as well as other source specific settings such as system prompt toggle.

Fixed

> 3. The list of models doesn't reflect the list of models supported in the Express mode. There're only these four: gemini-2.0-flash-001, gemini-2.0-flash-lite-001, gemini-2.5-pro-preview-03-25, gemini-2.5-flash-preview-04-17. They do not offer any gemma or gemini 1.5 models. And likely won't.

Fixed. I have tested all the currently remaining models, and they are all working.

> 4. What about features like tool calling and image inlining? Both would not work currently.

Fixed. Although Express mode doesn't support Gemini image generation model.

> 5. The code has changes to the caption endpoint, but doesn't have relevant changes to the caption extension. Captioning can be left out for now.

Todo

> 6. Current code wouldn't read `vertexai_model` field from presets.

Fixed

Cohee1207 committed:
Merge branch 'staging' into vertexai

Cohee1207 committed:
Log API name in the console

Cohee1207 committed:
Merge sysprompt toggles back

Cohee1207 committed:
Use Gemma tokenizers for Vertex and LearnLM

Cohee1207 committed:
AI Studio parity updates

Cohee1207 committed:
Add link to express mode doc. Also technically it's not a form

Cohee1207 committed:
Split title

Cohee1207 committed:
Use array includes

Cohee1207 reviewed (approved):
Ok, I managed to get a key and tested + fixed some stuff in the code. Waiting for resolution about captioning, then can merge.

NijikaMyWaifu committed:
Merge branch 'staging' into vertexai

NijikaMyWaifu committed:
Add support for Google Vertex AI in image captioning feature

Cohee1207 committed:
Specify caption API name, add to compression list
